task/amazon/resources/resource_security_group.go

```diff
@@ -0,0 +1,206 @@
+package resources
+
+import (
+	"context"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	"github.com/aws/aws-sdk-go-v2/service/ec2/types"
+
+	"terraform-provider-iterative/task/amazon/client"
+	"terraform-provider-iterative/task/universal"
+)
+
+func NewSecurityGroup(client *client.Client, identifier string, defaultVPC *DefaultVPC, firewall universal.Firewall) *SecurityGroup {
+	s := new(SecurityGroup)
+	s.Client = client
+	s.Identifier = universal.NormalizeIdentifier(identifier, true)
+	s.Attributes = firewall
+	s.Dependencies.DefaultVPC = defaultVPC
+	return s
+}
+
+type SecurityGroup struct {
+	Client       *client.Client
+	Identifier   string
+	Attributes   universal.Firewall
+	Dependencies struct {
+		*DefaultVPC
+	}
+	Resource *types.SecurityGroup
+}
+
+func (s *SecurityGroup) Create(ctx context.Context) error {
+	switch err := s.Read(ctx); {
+	case err == nil:
```

Comment:
The result of `s.Read(ctx)` can be `<nil>` if the resource already exists. See also idempotent operations and the awesome feature of importing existing resources after a failed creation operation.

Reply:
????

```go
err := s.Read(ctx); {
	if err != universal.NotFoundError {
		return err
	}
```

Its equivalent, because you are returning nil when is nil

Reply:
No, it isn't.

**Desired/current behavior**

```go
err := s.Read(ctx)
if err != universal.NotFoundError {
  return err # API ERROR
} else if err == nil {
  return nil # ALREADY EXISTS, NO NEED TO CREATE
} else {
  ··· # CREATE RESOURCE
}

Reply:
Maybe I got something wrong? 🤔

Reply:
universal.NotFoundError !== nil right?

```
if err != universal.NotFoundError {
  return err # API ERROR
} else if err == nil {
you are never entering here
```

if err is nil and you return nil is equivalent to return error which is nil so can be done with

if err != universal.NotFoundError {
  return err # API ERROR
}

where error can be something not universal.NotFoundError, including nil (that you return explicity)

Reply:
I do not propose an else

Reply:
I propose
```
err := s.Read(ctx); {
	if err != universal.NotFoundError {
		return err
	}

createInput := ec2.CreateSecurityGroupInput{
```

Reply:
I've edited https://github.com/iterative/terraform-provider-iterative/pull/237#discussion_r746975473 because it contained a major blunder. Please read it again.

Also removed `else` so you don't suffer. 😄

Reply:
Helio
```
if err == nil {
  return nil # ALREADY EXISTS, NO NEED TO CREATE
} else if err != universal.NotFoundError {
  return err # API ERROR
}
```
is equivalent to

```
 if err != universal.NotFoundError {
  return err # API ERROR
}
```

Reply:
step by step

```
if err == nil {
  return nil # ALREADY EXISTS, NO NEED TO CREATE
}

is equivalent to

if err == nil {
  return err // err is nil!
}
```

```
if err != universal.NotFoundError {
  return err # API ERROR
}
```

for err = nil returns nil

Reply:
🤦🏼‍♂️ What was I thinking on!? Thanks for having the patience of explaining it again.

Reply:
Update: I was thinking of `Delete` and not of `Create` 🤦🏼‍♂️